ruby/src/api.cc

```diff
@@ -0,0 +1,529 @@
+//    {
+//        if (!val) {
+//            return _nil;
+//        }
+//        if (auto ptr = dynamic_cast<ruby_value const*>(val)) {
+//            return ptr->value();
+//        }
+//        if (auto ptr = dynamic_cast<string_value const*>(val)) {
+//            return utf8_value(ptr->value());
+//        }
+//        if (auto ptr = dynamic_cast<integer_value const*>(val)) {
+//            return rb_int2inum(static_cast<SIGNED_VALUE>(ptr->value()));
+//        }
+//        if (auto ptr = dynamic_cast<boolean_value const*>(val)) {
+//            return ptr->value() ? _true : _false;
+//        }
+//        if (auto ptr = dynamic_cast<double_value const*>(val)) {
+//            return rb_float_new_in_heap(ptr->value());
+//        }
+//        if (auto ptr = dynamic_cast<array_value const*>(val)) {
+//            volatile VALUE array = rb_ary_new_capa(static_cast<long>(ptr->size()));
+//            ptr->each([&](value const* element) {
+//                rb_ary_push(array, to_ruby(element));
+//                return true;
+//            });
+//            return array;
+//        }
+//        if (auto ptr = dynamic_cast<map_value const*>(val)) {
+//            volatile VALUE hash = rb_hash_new();
+//            ptr->each([&](string const& name, value const* element) {
+//                rb_hash_aset(hash, utf8_value(name), to_ruby(element));
+//                return true;
+//            });
+//            return hash;
+//        }
+//        return _nil;
+//    }
+
+    VALUE api::lookup(std::initializer_list<std::string> const& names) const
+    {
+        volatile VALUE current = *rb_cObject;
+
+        for (auto const& name : names) {
+            current = rb_const_get(current, rb_intern(name.c_str()));
+        }
+        return current;
+    }
+
+    bool api::equals(VALUE first, VALUE second) const
+    {
+        return is_true(rb_funcall(first, rb_intern("eql?"), 1, second));
+    }
+
+    bool api::case_equals(VALUE first, VALUE second) const
+    {
+        return is_true(rb_funcall(first, rb_intern("==="), 1, second));
+    }
+
+    void api::eval(const string& code)
+    {
+        std::string exception;
+
+        rescue(
+                [&]() {
+                    rb_eval_string(code.c_str());
+                    return nil_value();
+                },
+                [&](VALUE exc) {
+                    exception = exception_to_string(exc);
+                    return nil_value();
+                });
+
+        if (!exception.empty()) {
+            throw runtime_error(exception);
+        }
+    }
+
+    lth_lib::dynamic_library api::find_library()
+    {
+        // First search for an already loaded Ruby.
+        auto library = find_loaded_library();
+        if (library.loaded()) {
+            return library;
+        }
+
+#ifdef LEATHERMAN_RUBY
+        // Ruby lib location was specified at compile-time, fix to that.
+        if (library.load(LEATHERMAN_RUBY)) {
```

Comment:
So yeah, add a static member to the class that defaults to an empty string, and instead of `#ifdef` do if string is not empty.